rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <chrono>
+#include <cstdio>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rosbag2_compression/sequential_compression_reader.hpp"
+#include "rosbag2_compression/sequential_compression_writer.hpp"
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+namespace rosbag2_py {
+
+class Reader {
+public:
+  Reader(const std::string &reader_class) {
+    if (!reader_class.compare("SequentialReader")) {
+      auto sequential_reader =
+          std::make_unique<rosbag2_cpp::readers::SequentialReader>();
+      reader_ =
+          std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else if (!reader_class.compare("SequentialCompressionReader")) {
+      auto sequential_reader =
+          std::make_unique<rosbag2_compression::SequentialCompressionReader>();
+      reader_ =
+          std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else {
+      throw std::runtime_error{"Reader class type " + reader_class +
+                               " not supported."};
+    }
+  }
+
+  void open(rosbag2_cpp::StorageOptions &storage_options,
+            rosbag2_cpp::ConverterOptions &converter_options) {
+    reader_->open(storage_options, converter_options);
+  }
+
+  bool has_next() { return reader_->has_next(); }
+
+  /// Return a tuple containing the topic name, the serialized ROS message, and
+  /// the timestamp.
+  pybind11::tuple read_next() {
+    const auto next = reader_->read_next();
+    rcutils_uint8_array_t rcutils_data = *next->serialized_data.get();
+    std::string serialized_data(
+        rcutils_data.buffer, rcutils_data.buffer + rcutils_data.buffer_length);
+    return pybind11::make_tuple(
+        next->topic_name, pybind11::bytes(serialized_data), next->time_stamp);
```

Comment:
It would be nice for `write()` to take a tuple of 3 arguments, parallel to what `read_next()` returns. Then the Python script wouldn't need access to a `SerializedBagMessage` type. With the current writer test script, the call would look like `writer.write('chatter', serialize_message(msg), time_stamp)`. All the work with `SerializedBagMessage` can then be done within `_rosbag2_py.cpp`, and then hopefully you wouldn't need to wrap `SerializedBagMessage`.

Reply:
Great suggestion, makes much more sense, I'll try it out :)  Thanks for helping out!

Reply:
Yay it worked :) Thanks! 

Let me know what you think is best for this PR: as described below, create a PR to your fork or just finish up this one as a separate one for a proper review now that it's working?